contracts/test/fixture/_pools.js

```diff
@@ -0,0 +1,158 @@
+
+    await this.balancerVault.connect(sAttacker).joinPool(
+      this.poolId,
+      sAttacker.address, // sender
+      sAttacker.address, // recipient
+      [
+        //JoinPoolRequest
+        this.assetAddressArray, // assets
+        maxAmountsIn, // maxAmountsIn
+        userData, // userData
+        false, // fromInternalBalance
+      ]
+    );
+
+    const expectedAmount = this.preTiltTVL.add(amount);
+    await this.verifyTiltAmounts(expectedAmount);
+  }
+
+  async untiltPool(sAttacker, attackingAsset) {
+    // for composableStable pools do not encode the BPT token in the user data request
+    const assetIndexAdjustement = this.poolType == "composableStable" ? -1 : 0;
+
+    /* encode user data for pool joining
+     *
+     * EXACT_BPT_IN_FOR_ONE_TOKEN_OUT:
+     * User sends a precise quantity of BPT, and receives an estimated
+     * but unknown (computed at run time) quantity of a single token
+     *
+     * ['uint256', 'uint256', 'uint256']
+     * [EXACT_BPT_IN_FOR_ONE_TOKEN_OUT, bptAmountIn, exitTokenIndex]
+     */
+    const userData = ethers.utils.defaultAbiCoder.encode(
+      ["uint256", "uint256", "uint256"],
+      [
+        0,
+        await this.bptToken.balanceOf(sAttacker.address),
+        BigNumber.from(
+          (
+            (await this.getAssetIndex(attackingAsset)) + assetIndexAdjustement
+          ).toString()
+        ),
+      ]
+    );
+
+    await this.bptToken
+      .connect(sAttacker)
+      .approve(this.balancerVault.address, oethUnits("1").mul(oethUnits("1"))); // 1e36
+
+    await this.balancerVault.connect(sAttacker).exitPool(
+      this.poolId,
+      sAttacker.address, // sender
+      sAttacker.address, // recipient
+      [
+        //ExitPoolRequest
+        this.assetAddressArray, // assets
+        Array(this.assetAddressArray.length).fill(BigNumber.from("0")), // minAmountsOut
+        userData, // userData
+        false, // fromInternalBalance
+      ]
+    );
+
+    if (this.preTiltTVL) {
+      await this.verifyTiltAmounts(this.preTiltTVL);
+    }
+  }
+
+  async getAssetIndex(asset) {
+    const tokens = (await this.balancerVault.getPoolTokens(this.poolId)).tokens;
+    for (let i = 0; i < tokens.length; i++) {
+      if (tokens[i].toLowerCase() === asset.address.toLowerCase()) {
+        return i;
+      }
+    }
+
+    throw new Error(
+      `Can not find the asset with address ${asset.address} in the pool id: ${this.poolId}`
+    );
+  }
+
+  async getTvl() {
+    if (this.poolType == "composableStable") {
+      return await this.bptToken.getActualSupply();
+    } else {
+      return await this.bptToken.totalSupply();
+    }
+  }
+}
+
+async function factoryCreatePool(config) {
+  if (config.platform === "balancer") {
+    return new BalancerPool(
+      config.poolId,
+      config.poolType,
+      config.assetAddressArray,
+      config.bptToken,
+      config.balancerVault
+    );
+  }
+  // else if platform === "curve"...
```

Comment:
nit: probably a good idea to throw an error here, would be easy to figure out what's wrong when this happens